optimize integration testing in CI

nashif opened the pull request:
This adds a new option, integration which is used to list platforms to use in integration testing (CI) and avoids whitelisting platforms. This will enable running those tests on real hardware if the --integration option is omitted.
--integration option is now not default, in the future this will be the default and will build/test a minimal set, running sanitycheck with more coverage will be possible use another option.

oh, one more thing in this PR: use native_posix exclusively where it makes sense and avoid massive build/running trivial stuff that is being tested elsewhere anyways..

nashif committed:
sanitycheck: add integration testing platforms

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: realign schema

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: introduce integration testing

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: philosophers: enable for integration testing

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: nfc: enable for integration testing

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: cmsis_dsp: use integration keyword

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: hello_world: filter for integration

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: cpp_synchronization: filter for integration

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
ci: enable sanitycheck for integration

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

galak reviewed (commented):
How about the YAML being `integration_board`?  The mix in sanitycheck of platform and our use of board everywhere else has always been a little funny.

nashif commented:
stats:

current master:
```
i9:zephyr(master): sanitycheck -y
Renaming output directory to /home/nashif/Work/zephyrproject/zephyr/sanity-out.5
INFO    - JOBS: 20
INFO    - Selecting default platforms per test case
INFO    - Building initial testcase list...
INFO    - 4206 test configurations selected, 202630 configurations discarded due to filters.
INFO    - Completed in 3 seconds
```

this PR with no options:
```
i9:zephyr(sanitycheck/optimize_tests): sanitycheck -y
Renaming output directory to /home/nashif/Work/zephyrproject/zephyr/sanity-out.3
INFO    - JOBS: 20
INFO    - Selecting default platforms per test case
INFO    - Building initial testcase list...
INFO    - 4444 test configurations selected, 202392 configurations discarded due to filters.
INFO    - Completed in 3 seconds
```

this PR with -G
```
i9:zephyr(sanitycheck/optimize_tests): sanitycheck -y -G
Renaming output directory to /home/nashif/Work/zephyrproject/zephyr/sanity-out.4
INFO    - JOBS: 20
INFO    - Selecting default platforms per test case
INFO    - Building initial testcase list...
INFO    - 4033 test configurations selected, 202803 configurations discarded due to filters.
INFO    - Completed in 3 seconds
```

nashif committed:
tests: dac: remove whitelist

We already depend on dac.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: benchmarks: filter for integration

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tinycrypt

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
cleanup prj*.conf

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
drivers: arcv2_irq_unit: remove include

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
power: remove unused inclusion of soc.h

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: device: remove whitelisting

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: shell: remove filter

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests/samples: enable for integration testing

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: support pre-filter

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: pi: add integration support

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samepls/tests: use pre-filter

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: pwm_servo: use dt_alias_exists

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
boards: qemu_riscv: disable spi in DTS

nashif committed:
boards: intel_apl_adsp: we do not use mcuboot

nashif committed:
boards: qemu_x86_nommu/coverage: do not run on drivers/samples

nashif committed:
sanitycheck: add subsys_dt_compat_enabled_and_match

nashif committed:
remove more depends_on

nashif committed:
samples: basic: remove FIXME

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: display: fix filters

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
ci: create dts_db for pre-filter

This should go into sanitycheck directly in the future and should not be
needed to be run before.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

erwango commented on samples/sensor/lps22hb/sample.yaml:
```diff
@@ -4,7 +4,8 @@ tests:
     harness: console
     tags: sensors
-    depends_on: i2c lps22hb
+    pre-filter: subsys_dt_compat_enabled("i2c")
```
I wonder what is the difference between this and the above change:
```
- depends_on: i2c lis2dh
+ pre-filter: subsys_dt_compat_enabled("i2c") and dt_compat_enabled("st,lis2dh")
```

nashif replied:
the above change probably need to be changed, i2c in pre-filter, and lis2dh in filter.

erwango commented:
@nashif do you think it would be feasible to override integration boards locally ? (I'm thinking about a new prop in map.yaml)

nashif commented:
> @nashif do you think it would be feasible to override integration boards locally ? (I'm thinking about a new prop in map.yaml)

why would you want to do that? you can still run on any boards with full coverage by specifying the boards directly.

erwango commented:
> why would you want to do that? you can still run on any boards with full coverage by specifying the boards directly.

Maybe I'm not quite clear with the new option. And reading again the description, I'm actually more confused. I understand the interest of removing the withelisting, but I'm not clear why the integration makes things better.
Can you provide more details on the problematic you're tackling here ?

nashif commented:
> Can you provide more details on the problematic you're tackling here ?

--integration helps optimize the coverage for CI runs, everything else is kept the same or is improved by relying on up2date information from dts re hardware availability and support.

The goal is not to randomly build and run tests and instead do this in the most optimal way balance coverage with run-time required in CI when someone submits a PR.

erwango commented:
> The goal is not to randomly build and run tests and instead do this in the most optimal way balance coverage with run-time required in CI when someone submits a PR.

Get it better, txs

nvlsianpu commented:
Did I understand this patch well: apart form what listed under integration, qemu_x86 will be always one of  CI target as well (if filters mach?).
Currently settings, storage and FS are tested in CI on either qemu and posix (which is great, and i really like as extensive testing are problematic for flash endurance). I wan't to avoid disabling these test in CI accidentally.

BTW: some of commits here are helpers patches, but not strongly connected and can go to the master independently, even right now.

tyrantlizard77 reviewed (commented):
can you come up with something shorter than `subsys_dt_compat_enabled` for those of us who don't like to type? literally every `pre-filer:` statement uses it

nashif commented:
superseded with #28435
